File: app/src/main/java/pl/softfly/flashcards/dao/app/DeckDao.java
package pl.softfly.flashcards.dao.app;

import androidx.annotation.NonNull;
import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.Update;

import io.reactivex.rxjava3.core.Maybe;
import io.reactivex.rxjava3.schedulers.Schedulers;
import pl.softfly.flashcards.db.TimeUtil;
import pl.softfly.flashcards.entity.app.Deck;

/**
 * Naming convention:
 * https://docs.spring.io/spring-data/jpa/docs/2.5.5/reference/html/#jpa.query-methods.query-creation
 */
@Dao
public abstract class DeckDao {

    @Query("SELECT * FROM Deck WHERE path=:path")
    public abstract Deck findByKey(String path);

    @NonNull
    @Insert
    public abstract long insert(Deck deck);

    @NonNull
    @Update
    public abstract void update(Deck deck);

    public Deck refreshLastUpdatedAt(String path) {
        if (!path.endsWith(".db")) {
            path += ".db";
        }

        Deck deck = findByKey(path);
        if (deck != null) {
            deck.setLastUpdatedAt(TimeUtil.getNowEpochSec());
            update(deck);
        } else {
            deck = new Deck();
            deck.setName(getDeckName(path));
            deck.setPath(path);
            deck.setLastUpdatedAt(TimeUtil.getNowEpochSec());
            insert(deck);
        }
        return deck;
    }

    @NonNull
    protected String getDeckName(@NonNull String deckDbPath) {
        return deckDbPath.substring(deckDbPath.lastIndexOf("/") + 1, deckDbPath.length() - 3);
    }

}

File: room/src/main/java/pl/gocards/room/dao/app/DeckDao.kt
package pl.gocards.room.dao.app

import androidx.room.Dao
import androidx.room.Query
import androidx.room.Transaction
import pl.gocards.room.dao.BaseDao
import pl.gocards.room.entity.app.Deck
import pl.gocards.room.util.DbUtil
import pl.gocards.room.util.TimeUtil

/**
 * Naming convention:
 * https://docs.spring.io/spring-data/jpa/docs/2.5.5/reference/html/#jpa.query-methods.query-creation
 *
 * @author Grzegorz Ziemski
 */
@Dao
abstract class DeckDao : BaseDao<Deck> {

    @Query("SELECT * FROM Deck WHERE path=:path")
    protected abstract fun findByKey(path: String): Deck?

    @Query("DELETE FROM Deck WHERE `path` LIKE '%' || :path || '%'")
    abstract fun deleteByStartWithPath(path: String)

    @Transaction
    open fun refreshLastUpdatedAt(path: String): Deck {
        val pathWithDb = DbUtil.addDbExtension(path)
        var deck = findByKey(pathWithDb)
        if (deck != null) {
            deck.lastUpdatedAt = TimeUtil.getNowEpochSec()
            update(deck)
        } else {
            deck = Deck(
                name = DbUtil.getDeckName(pathWithDb),
                path = pathWithDb,
                lastUpdatedAt = TimeUtil.getNowEpochSec()
            )
            insert(deck)
        }
        return deck
    }
}